### ISSUE TITLE
subtraction overflow panic in ListView::required_size

### ISSUE THREAD

**Issue by scottlamb** [gyscos/Cursive#207]
Bug report.

# Problem description

This dialog I've been working on:

```rust
extern crate cursive;

use cursive::Cursive;
use cursive::traits::Boxable;
use cursive::views;

fn main() {
    let mut siv = Cursive::new();
    let mut layout = views::LinearLayout::vertical()
        .child(views::ListView::new()
            .child("id", views::TextView::new("<new>"))
            .child("uuid", views::TextView::new("<new>"))
            .child("short name", views::EditView::new())
            .child("host", views::EditView::new())
            .child("username", views::EditView::new())
            .child("password", views::EditView::new())
            .min_height(6))
        .child(views::TextView::new("description"))
        .child(views::TextArea::new().min_height(3));
    for t in &["main", "sub"] {
        let list = views::ListView::new()
            .child("rtsp path", views::LinearLayout::horizontal()
                .child(views::EditView::new().full_width())
                .child(views::DummyView)
                .child(views::Button::new("Test", |_| ())))
            .child("sample file dir", views::SelectView::new()
                   .item("<none>", None as Option<i32>)
                   .popup())
            .child("record", views::Checkbox::new())
            .child("usage/capacity", views::TextView::new("<new>"))
            .min_height(4);
        layout.add_child(views::DummyView);
        layout.add_child(views::TextView::new(format!("{} stream", t)));
        layout.add_child(list);
    }
    siv.add_layer(
        views::Dialog::around(layout)
        .title("foo")
        .button("Add", |_| ())
        .dismiss_button("Cancel"));
    siv.run();
```

crashes if you open it in an 80x24 window on a debug build and scroll down so the "sub stream" portion is visible. The panic looks like this:

```
thread 'main' panicked at 'attempt to subtract with overflow', /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/list_view.rs:375:21
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:68
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:57
             at src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:397
   4: std::panicking::begin_panic
             at src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at src/libstd/panicking.rs:538
   6: std::panicking::try::do_call
             at src/libstd/panicking.rs:522
   7: std::panicking::try::do_call
             at src/libstd/panicking.rs:498
   8: <core::ops::range::Range<Idx> as core::fmt::Debug>::fmt
             at src/libcore/panicking.rs:71
   9: <core::ops::range::Range<Idx> as core::fmt::Debug>::fmt
             at src/libcore/panicking.rs:51
  10: <cursive::views::list_view::ListView as cursive::view::View>::required_size::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/list_view.rs:375
  11: cursive::view::view_wrapper::ViewWrapper::wrap_on_event::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  12: <cursive::views::box_view::BoxView<T> as cursive::view::view_wrapper::ViewWrapper>::with_view_mut
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:177
  13: cursive::view::view_wrapper::ViewWrapper::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  14: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  15: cursive::view::view_wrapper::ViewWrapper::wrap_on_event::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  16: <T as cursive::view::view_wrapper::ViewWrapper>::with_view_mut
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:101
  17: cursive::view::view_wrapper::ViewWrapper::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  18: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  19: <cursive::views::linear_layout::LinearLayout as cursive::view::View>::take_focus::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/linear_layout.rs:495
  20: cursive::view::view_wrapper::ViewWrapper::wrap_on_event::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  21: <T as cursive::view::view_wrapper::ViewWrapper>::with_view_mut
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:101
  22: cursive::view::view_wrapper::ViewWrapper::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  23: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  24: cursive::view::view_wrapper::ViewWrapper::wrap_on_event::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  25: <cursive::views::sized_view::SizedView<T> as cursive::view::view_wrapper::ViewWrapper>::with_view_mut
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:177
  26: cursive::view::view_wrapper::ViewWrapper::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  27: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  28: cursive::views::dialog::Dialog::on_event_content
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/dialog.rs:208
  29: cursive::views::dialog::Dialog::check_focus_grab::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/dialog.rs:489
  30: cursive::view::view_wrapper::ViewWrapper::wrap_on_event::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  31: <cursive::views::layer::Layer<T> as cursive::view::view_wrapper::ViewWrapper>::with_view_mut
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:177
  32: cursive::view::view_wrapper::ViewWrapper::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  33: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  34: <cursive::views::shadow_view::ShadowView<T> as cursive::view::view_wrapper::ViewWrapper>::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/shadow_view.rs:67
  35: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  36: cursive::view::view_wrapper::ViewWrapper::wrap_on_event::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  37: <T as cursive::view::view_wrapper::ViewWrapper>::with_view_mut
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:101
  38: cursive::view::view_wrapper::ViewWrapper::wrap_on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:50
  39: cursive::view::view_wrapper::<impl cursive::view::View for T>::on_event
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/view/view_wrapper.rs:116
  40: <cursive::views::stack_view::StackView as cursive::view::View>::draw::{{closure}}
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/views/stack_view.rs:223
  41: cursive::cursive::Cursive::draw
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/cursive.rs:541
  42: cursive::cursive::Cursive::draw
             at /Users/slamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cursive-0.7.5/src/cursive.rs:484
  43: crash::main
             at src/main.rs:41
  44: panic_unwind::dwarf::eh::read_encoded_pointer
             at src/libpanic_unwind/lib.rs:101
  45: rust_panic
             at src/libstd/panicking.rs:459
             at src/libstd/panic.rs:365
             at src/libstd/rt.rs:58
  46: crash::main::{{closure}}
```

# Environment

* Operating system: macOS
* Backend used: ncurses 5.7
* Current locale (run `locale` in a terminal). en_US.UTF-8
* Cursive version (from crates.io, from git, ...): 0.7.5

**Comment by gyscos** [comment]
Indeed, another case when we're bitten by unchecked subtraction. Thanks for the report!

It should be fixed in 8f3441bc3c0b1081ec79e099f9fff8fb2eac8f7a.

**Comment by gyscos** [comment]
The fix is now part of 0.8.2.
